### PR TITLE
RFC: Add feature mapping to BLR

### PR DESCRIPTION
**Pull request by rossviljoen**
Adds a feature mapping function to the definition of BLR

Really just a convenience that saves one from having to carry around the feature mapping separately and call `blr(phi(x))` each time (like in the `nn-blr` example in this repo)

@willtebbutt do you think the feature mapping belongs here or no? If so, I'll write tests and docs etc.

One problem with this is that we'd have to reimplement matrix obsdim style input here, which is some ugly code duplication

### CODE REVIEW

**Comment by willtebbutt** [comment]
Hmmm to be honest I would be more inclined towards introducing a new type called `BasisFunctionBayesianLinearRegressor`, or something, that wraps a linear regressor and phi. I'd put it in a separate file too.

> One problem with this is that we'd have to reimplement matrix obsdim style input here, which is some ugly code duplication

Could you expand on this a bit? I'm not seeing what you're getting at.

**Comment by rossviljoen** [comment]
> Hmmm to be honest I would be more inclined towards introducing a new type called BasisFunctionBayesianLinearRegressor, or something, that wraps a linear regressor and phi. I'd put it in a separate file too.

Sounds good, that was the other option

>Could you expand on this a bit? I'm not seeing what you're getting at.

Yes, it turns out I'm just a bit thick - not needed at all

**Comment by rossviljoen** [comment]
This has turned into a much bigger PR because I also moved function sampling into its own file - I could split that out into another if you want?

File: src/basis_function_regression.jl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+struct BasisFunctionBayesianLinearRegressor{Tblr<:BayesianLinearRegressor,Tϕ} <: AbstractGP
```
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/basis_function_regression.jl:1]: Sounds good to me.

[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/basis_function_regression.jl:1]: Also, this needs a docstring explaining the properties that phi should have.

**Comment by willtebbutt** [comment]
This broadly seems good. Just need to sort out the coverage drop, add some docs (docstring + a bit in the readme), and a patch bump.

**Comment by rossviljoen** [comment]
@willtebbutt have added docstrings and readme update.

**Review by willtebbutt — APPROVED** [review]
LGTM! Please merge + tag :)

**Comment by mvsoom** [comment]
This is great. :)

**Pull request merged**